### PR TITLE
add "/shoutout" command

### PR DESCRIPTION
**Pull request by olafyang**
# Description

This pr adds a new "/shoutout" command which calls the helix api and shouts out the specified user.

Add:
- `/shoutout <username>` calls `chat/shoutouts` of helix api and gives specified user a twitch shoutout.

It seems like twitch is returning a 500 for all errors excluding 429 rate limit . Trying to shoutout an user who's not a mod gives back to following in the log. Therefore I can not provide more detailed error message for now.

```
chatterino.twitch: Helix send shoutout, unhandled error data: 500 "{\"error\":\"Internal Server Error\",\"status\":500,\"message\":\"\"}" QJsonObject({"error":"Internal Server Error","message":"","status":500})
```

### CODE REVIEW

File: src/providers/twitch/api/Helix.hpp [hunk 1]
```diff
@@ -1318,6 +1333,12 @@ class Helix final : public IHelix
                           FailureCallback<HelixUpdateShieldModeError, QString>
                               failureCallback) final;
 
+    // https://dev.twitch.tv/docs/api/reference/#send-a-shoutout
+    void sendShoutout(
+        QString fromBroadCasterID, QString toBroadcasterID, QString moderatorID,
```
Comment (Felanbird) [inline src/providers/twitch/api/Helix.hpp:1338]:
```suggestion
        QString fromBroadcasterID, QString toBroadcasterID, QString moderatorID,
```

File: tests/src/HighlightController.cpp [hunk 2]
```diff
@@ -434,6 +434,14 @@ class MockHelix : public IHelix
                       failureCallback)),
                 (override));
 
+    // /shoutout
+    MOCK_METHOD(
+        void, sendShoutout,
+        (QString fromBroadCasterID, QString toBroadcasterID,
```
Comment (Felanbird) [inline tests/src/HighlightController.cpp:440]:
```suggestion
        (QString fromBroadcasterID, QString toBroadcasterID,
```

File: src/providers/twitch/api/Helix.cpp [hunk 3]
```diff
@@ -2631,6 +2631,66 @@ void Helix::updateShieldMode(
         .execute();
 }
 
+// https://dev.twitch.tv/docs/api/reference/#send-a-shoutout
+void Helix::sendShoutout(
+    QString fromBroadCasterID, QString toBroadcasterID, QString moderatorID,
```
Comment (Felanbird) [inline src/providers/twitch/api/Helix.cpp:2636]:
```suggestion
    QString fromBroadcasterID, QString toBroadcasterID, QString moderatorID,
```

File: src/providers/twitch/api/Helix.cpp [hunk 4]
```diff
@@ -2631,6 +2631,66 @@ void Helix::updateShieldMode(
...
+    ResultCallback<> successCallback,
+    FailureCallback<HelixSendShoutOutError, QString> failureCallback)
+{
+    using Error = HelixSendShoutOutError;
+
+    QUrlQuery urlQuery;
+    urlQuery.addQueryItem("from_broadcaster_id", fromBroadCasterID);
```
Comment (Felanbird) [inline src/providers/twitch/api/Helix.cpp:2643]:
```suggestion
    urlQuery.addQueryItem("from_broadcaster_id", fromBroadcasterID);
```

**Comment by Felanbird** [comment]
> It seems like twitch is returning a 500 for all errors excluding 429 rate limit 

I'm only able to replicate this issue when the shoutout sender is not a mod, I receive back a correct 401 when the `moderator_id` does not match the relevant oauth token, as well as a 400 when `from_broadcaster_id` is not live/has no viewers.

**Review by Felanbird — CHANGES_REQUESTED** [review]

**Comment by olafyang** [comment]
> > It seems like twitch is returning a 500 for all errors excluding 429 rate limit
> 
> I'm only able to replicate this issue when the shoutout sender is not a mod, I receive back a correct 401 when the `moderator_id` does not match the relevant oauth token, as well as a 400 when `from_broadcaster_id` is not live/has no viewers.
> 
> [twitchdev/issues#779](https://github.com/twitchdev/issues/issues/779)

i'll check on my end too and make changes, thanks!!

**Comment by olafyang** [comment]
Typos are fixed in e52f8bba6e6b20785b63c4ced287059aa4f7d736.

I've also added some error handling for the api.

File: src/controllers/commands/CommandController.cpp [hunk 5]
```diff
@@ -3211,6 +3211,111 @@ void CommandController::initialize(Settings &, Paths &paths)
...
+                                    "You are being ratelimited by Twitch. "
+                                    "Try again in a few seconds.";
+                            }
+                            break;
+
+                            case Error::UserNotMod: {
+                                errorMessage +=
+                                    "You are not a moderator of this channel.";
+                            }
+                            break;
```
Comment (pajlada) [inline src/controllers/commands/CommandController.cpp:3287]: I would think the "UserNotAuthorized" error covers this case already, so just remove the UserNotMod error case & make it UserNotAuthorized instead

File: src/providers/twitch/api/Helix.cpp [hunk 6]
```diff
@@ -2631,6 +2631,96 @@ void Helix::updateShieldMode(
...
+                }
+                case 429: {
+                    failureCallback(Error::Ratelimited, message);
+                    break;
+                }
+                    // Helix returns 500 when user is not mod
+                case 500: {
+                    failureCallback(Error::UserNotMod,
+                                    "You are not a moderator.");
+                    break;
```
Comment (pajlada) [inline src/providers/twitch/api/Helix.cpp:2711]: I would prefer to treat 500s as unhandled errors - just forward the message we get from twitch

File: src/controllers/commands/CommandController.cpp [hunk 7]
```diff
@@ -3211,6 +3211,111 @@ void CommandController::initialize(Settings &, Paths &paths)
...
+                                    "The broadcaster is not streaming live or "
+                                    "does not have one or more viewers.";
+                            }
+                            break;
+
+                            case Error::Unknown: {
+                                errorMessage +=
+                                    "An unknown error has occurred.";
+                            }
+                            break;
```
Comment (pajlada) [inline src/controllers/commands/CommandController.cpp:3306]: I would be ok with including the `message` portion in unhandled errors here, so the user would see something like "Internal server error" or w/e if we don't handle one of the error cases.

File: src/controllers/commands/CommandController.cpp [hunk 8]
```diff
@@ -3211,6 +3211,111 @@ void CommandController::initialize(Settings &, Paths &paths)
 
     this->registerCommand("/shield", &commands::shieldModeOn);
     this->registerCommand("/shieldoff", &commands::shieldModeOff);
+
+    this->registerCommand("/shoutout", [](const QStringList &words,
+                                          auto channel) {
```
Comment (pajlada) [inline src/controllers/commands/CommandController.cpp:3216]: I would prefer if this command was moved to its own file under `src/controllers/commands/builtin/twitch` (see `ShieldMode.hpp`/`ShieldMode.cpp`)

**Review by pajlada — CHANGES_REQUESTED** [review]
Looks good overall, some nitpicks with error handling & where the code is located

File: src/providers/twitch/api/Helix.cpp (hunk 6 above)
Comment (Mm2PL) [reply src/providers/twitch/api/Helix.cpp:2711]: Twitch likes to give an empty message on 500s so there needs to be a genetic "twitch fucked up" error case and one that forwards a non-empty message.

**Comment by olafyang** [comment]
oops i forgot to return at the anon check and the argument check, i'll fix them in the evening once i get home

**Comment by pajlada** [comment]
Currently fixing the merge conflicts, will solve the returns in the same batch

**Review by pajlada — APPROVED** [review]
Looks good to me, works as expected in the few scenarios I tested it

When you've taken a look at the changes @olafyang we can go ahead and merge this in

**Comment by olafyang** [comment]
yep looks fantastic!! please go ahead and merge it. i learned a lot from this, thanks ❤️❤️ would love to contribute some more in the future

File: src/providers/twitch/api/Helix.cpp [hunk 9]
```diff
@@ -2597,6 +2597,106 @@ void Helix::updateShieldMode(
...
+                case 429: {
+                    failureCallback(Error::Ratelimited, message);
+                }
+                break;
+
+                case 500: {
+                    // Helix returns 500 when user is not mod,
+                    if (message.isEmpty())
+                    {
+                        failureCallback(Error::Unknown, "Empty message");
```
Comment (Mm2PL) [inline src/providers/twitch/api/Helix.cpp:2680]: Can we maybe add a thing to the message that the api fails when not mod?

Comment (pajlada) [reply src/providers/twitch/api/Helix.cpp:2680]: There was originally, but having that error for a generic error 500 will lead to outdated or wrong errors.
I would rather incorrectly trust Twitch's error handling than spit out bad errors ourselves

**Review by Felanbird — APPROVED** [review]

File: src/providers/twitch/api/Helix.cpp (hunk 9 above)
Comment (pajlada) [reply src/providers/twitch/api/Helix.cpp]: message updated from "Empty message" to "Twitch internal server error" to give users a bit more of a clear idea of what's going on

**Comment by pajlada** [comment]
Thank you! As a first-time contributor, you can now add yourself to the contributors list that's shown inside the About page in Chatterino.

If you want this, you can open a new PR where you modify the `resources/contributors.txt` file and add yourself to the list. Make sure to read the comments at the top for instructions.

**Comment by Felanbird** [comment]
Also you'll want to do it on a branch after you clean up your master branch, as creating this PR from your master branch has made a mess for you 😁

**Pull request merged**